mobx/lib/src/api/observable_collections/observable_list.dart

```diff
@@ -259,69 +271,122 @@ class ObservableList<T>
   @override
   void removeRange(int start, int end) {
     _context.conditionallyRunInAction(() {
-      final removedItems = _list.sublist(start, end);
-      _list.removeRange(start, end);
-      _notifyListUpdate(start, null, removedItems);
+      if (end > start) {
+        final removedItems = _list.sublist(start, end);
+        _list.removeRange(start, end);
+        _notifyRangeUpdate(start, null, removedItems);
+      }
     }, _atom);
   }
 
   @override
   void removeWhere(bool Function(T element) test) {
     _context.conditionallyRunInAction(() {
-      final removedItems = _list.where(test).toList(growable: false);
-      _list.removeWhere(test);
-      _notifyListUpdate(0, null, removedItems);
+      final removedElements = <int, ElementChange>{};
+      for (var i = _list.length - 1; i >= 0; --i) {
+        final element = _list[i];
+        if (test(element)) {
+          removedElements[i] = ElementChange(oldValue: element, type: OperationType.remove);
+          _list.removeAt(i);
+        }
+      }
+      if (removedElements.isNotEmpty) {
+        _notifyElementsUpdate(removedElements);
+      }
     }, _atom);
   }
 
   @override
   void replaceRange(int start, int end, Iterable<T> newContents) {
     _context.conditionallyRunInAction(() {
-      final oldContents = _list.sublist(start, end);
-      _list.replaceRange(start, end, newContents);
-      _notifyListUpdate(start, newContents, oldContents);
+      if (end > start || newContents.isNotEmpty) {
+        final oldContents = _list.sublist(start, end);
+        _list.replaceRange(start, end, newContents);
+        _notifyRangeUpdate(start, newContents.toList(growable: false), oldContents);
+      }
     }, _atom);
   }
 
   @override
   void retainWhere(bool Function(T element) test) {
     _context.conditionallyRunInAction(() {
-      final removedItems = _list.where((_) => !test(_)).toList(growable: false);
-
-      _list.retainWhere(test);
-      _notifyListUpdate(0, null, removedItems);
+      final removedElements = <int, ElementChange>{};
```

Comment:
Rather than storing the `removedElements` as a Map, its probably better to store the `index` in the `ElementChange` itself as a field? That way the element-changes would be a list and easier to iterate ? Curious what you think.

Reply:
On second thought, using a Map keeps the `ElementChange` agnostic of the `index` itself and as a consumer you would iterate on the Map `keys` instead ?

Reply:
Thanks!
Actually using the Map ensures that change of the element with specific index will be unique over a single ListChange object. It must be unique, and the interface of ListChange class guaranties this. Probably there was no other serious reasons for me to move the `index` field outside of the ElementChange class.
The same idea applies to the RangeChange class except that there are no checks of ranges overlaying for now.
But you mentioned the sorting, and now I think you were right. The order does much more matter than uniqueness in interface and makes consumers more efficient.

Regarding the unit tests simplifying.. I'm not sure I correctly understand what you mean. Are you talking about using of checks (`expect`) inside of consumer closure body? If so, sure I'll fix these issues.

Reply:
Ya the `Map` makes more sense now.

Looks like the coverage has dropped. Can you bring it back to 99%?

Reply:
I think unique-ness will be guaranteed as in a single pass of the observe, you will not see simultaneous changes on the same index, assuming we are running this on a single thread!

In that case, have the index inside the ElementChange would be easier to handle. The unit tests looks more readable now, just that the coverage needs to be improved.

Reply:
It seems that codecov complains only on 2 lines with `assert(index != null)`. I can't realize how to cover these expressions by tests. Could you help me please? How do you usually do this?

Reply:
Ok, I figured it out. Check this out please.